blockchain/vm: vmLog/feePayer/validateSender precompiled contract address mapping

yoomee1313 opened the pull request:
## Proposed changes
- vmLog/feePayer/validateSender precompiled contracts' address will be changed after the istanbulCompatible change. You can see how the address changes by looking at the yellow highlights in the table below.
![image](https://user-images.githubusercontent.com/64360042/110870821-90622a80-8310-11eb-95e1-afd9f60a00ae.png)

- However, the vmLog/feePayer/validateSender calling functions that are already in use cannot be used after the change. By this PR, we can still use those functions by utilizing CodeFormat. It is used as an incompatible change version. 
- Originally, when a smart contract is released, CodeFormatEVM is assigned to that contract. By this PR, CodeFormatEVM2 is assigned to the new contracts which will be deployed after istanbulCompatible change.
- The result below shows the success result of calling precompiled contract according to those conditions: node's incompatible change version (constantinople/istanbul), deployment time, and call time. `before` means a contract is deployed or called before the IstanbulCompatible change. `after` means a contract is deployed or called after the IstanbulCompatible change.
![image](https://user-images.githubusercontent.com/64360042/112439444-756dcc80-8d8c-11eb-963b-3c61fbfb530d.png)

## Types of changes

Please put an x in the boxes related to your change.

- [ ] Bugfix
- [x] New feature or enhancement
- [ ] Others

## Checklist

*Put an x in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This is simply a reminder of what we are going to look for before merging your code.*

- [x] I have read the [CONTRIBUTING GUIDELINES](https://github.com/klaytn/klaytn/blob/master/CONTRIBUTING.md) doc
- [x] I have signed the [CLA](https://cla-assistant.io/klaytn/klaytn)
- [x] Lint and unit tests pass locally with my changes (`$ make test`)
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any dependent changes have been merged and published in downstream modules

## Related issues

- Please leave the issue numbers or links related to this PR here.

## Further comments

If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...

KimKyungup commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,33 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+		// (1) check whether the contract(caller) is deployed before istanbul incompatible change
+		// (2) then, check whether the precompiled contract(address) is 0x09, 0x0a, 0x0b
+		// (3) if it is, use old precompiled contract set
+		if evm.StateDB.GetCodeFormat(caller) == params.CodeFormatEVM &&
+			(address == common.HexToAddress("0x09") || address == common.HexToAddress("0x0a") || address == common.HexToAddress("0x0b")) {
```
@yoomee1313 Can this code be simplified like below?
```suggestion
		if evm.StateDB.GetCodeFormat(caller) == params.CodeFormatEVM {
```

yoomee1313 replied:
Gas Costs of bn256 precompiled contracts(0x06, 0x07, 0x08) will be changed in IstanbulCompatible Change. So, this code is necessary!

winnie-byun replied:
@yoomee1313 Could you explain more or make the next PR about how you are going to change with gas costs? I don't get how things will change.

yoomee1313 replied:
As we discussed at offline, 0x06, 0x07, 0x08 should use istanbul precompiled contract map even though the contract is deployed before the istanbulCompatible change. Only 0x09, 0x0a, 0x0b needs constantinople precompiled contract map if the contract is deployed before the change.

winnie-byun replied:
@yoomee1313 Thanks for explaining. I get it now :)

KimKyungup replied:
I think this is not a good idea. Let's discuss this in offline together.

aidan-kwon commented:
@yoomee1313 I don't understand why Istanbul Address is callable when deployment time is **Before** and call time is **After**. I think calling Istanbul Address of the contact may not be intended by the contract author.

aidan-kwon commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,33 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+	}
+}
+
+func (evm *EVM) GetPrecompiledContract(caller, address common.Address) map[common.Address]PrecompiledContract {
```
I think `GetPrecompiledContractMap` is more proper for a name.

yoomee1313 replied:
Thanks, it's a good suggestion. Reflected at https://github.com/klaytn/klaytn/pull/912/commits/1c0928ba549158af0d357eed0e6d0a7763e14bd8.

aidan-kwon commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,33 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+func (evm *EVM) GetPrecompiledContract(caller, address common.Address) map[common.Address]PrecompiledContract {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		// Below code is added to make vmLog(0x09), feePayer(0x0a), validateSender(0x0b) precompiled contract
```
```suggestion
		// The following code is added to make vmLog(0x09), feePayer(0x0a), validateSender(0x0b) precompiled contract
```

yoomee1313 replied:
Reflected at https://github.com/klaytn/klaytn/pull/912/commits/1c0928ba549158af0d357eed0e6d0a7763e14bd8.

aidan-kwon commented on params/protocol_params.go:
```diff
@@ -150,13 +150,16 @@ type CodeFormat uint8
 
 const (
 	CodeFormatEVM CodeFormat = iota
+	CodeFormatEVM2
```
We need more information to distinguish `CodeFormatEVM` and `CodeFormatEVM2`. How about adding comments or changing its name like  `CodeFormatEVMIstanbulCompatible`.

yoomee1313 replied:
I changed the name of CodeFormatEVM2 and added some description. Reflected at https://github.com/klaytn/klaytn/pull/912/commits/1c0928ba549158af0d357eed0e6d0a7763e14bd8.

yoomee1313 commented on blockchain/vm/contracts.go:
```diff
@@ -431,6 +431,7 @@ func (c *vmLog) GetRequiredGasAndComputationCost(input []byte) (uint64, uint64)
 
 // Runs the vmLog contract.
 func (c *vmLog) Run(input []byte, contract *Contract, evm *EVM) ([]byte, error) {
+	logger.Info("Precompiled Contract Run", "vmLog", strconv.QuoteToASCII(string(input)))
```
This is my mistake! I'll remove it.

yoomee1313 replied:
Reflected at https://github.com/klaytn/klaytn/pull/912/commits/1c0928ba549158af0d357eed0e6d0a7763e14bd8

winnie-byun commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,33 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
 }
 
+func (evm *EVM) CurrentCodeFormat() params.CodeFormat {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		return params.CodeFormatEVM2
+	default:
+		return params.CodeFormatEVM
+	}
```
Shouldn't this code be decided by the blocknumber?
```suggestion
	switch {
	case evm.chainRules.IsIstanbul:
	        switch block :=  evm.BlockNumber; {
	        case block < 50M:
	                return params.CodeFormatEVM
	        default:
	        	return params.CodeFormatEVM2
	        }
	default:
		return params.CodeFormatEVM2
	}
	return params.CodeFormatEVM2
```

yoomee1313 replied:
evm.chainRules.IsIstanbul does the same thing. It is set before executing according to the node's current block number. I'll add some comments at chainRules.

winnie-byun commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,33 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+	case evm.chainRules.IsIstanbul:
+		// Below code is added to make vmLog(0x09), feePayer(0x0a), validateSender(0x0b) precompiled contract
+		// works after the istanbul incompatible change
+		// (1) check whether the contract(caller) is deployed before istanbul incompatible change
+		// (2) then, check whether the precompiled contract(address) is 0x09, 0x0a, 0x0b
+		// (3) if it is, use old precompiled contract set
```
```suggestion
		// if contract(caller) is deployed before istanbul and 
		// (1) contract address is 0x09, 0x0a, 0x0b, use the old  precompiled contract set
		// (2) contract address is 0x06, 0x07, 0x08, use the new  precompiled contract set
		//       (the only difference is gas cost)
```

yoomee1313 replied:
Thanks for revising comments, it's really helpful. Reflected at https://github.com/klaytn/klaytn/pull/912/commits/1c0928ba549158af0d357eed0e6d0a7763e14bd8.

yoomee1313 commented:
@aidan-kwon
>  I don't understand why Istanbul Address is callable when deployment time is Before and call time is After.

I never thought deeply about what I pointed out. Thank you. I checked how geth handled it, and the Istanbul address is callable when the contract is deployed before the istanbul change. Unfortunately, I couldn't find the reason or related documentation about this. 

> I think calling Istanbul Address of the contact may not be intended by the contract author.

Since this is a new item address, it will not be the case that the author did not intend.

aidan-kwon commented on params/protocol_params.go:
```diff
@@ -149,14 +149,17 @@ func GetMaximumExtraDataSize() uint64 {
 	case CodeFormatEVM:
 		return "CodeFormatEVM"
+	case CodeFormatEVMIstanbulCompatible:
+		return "CodeFormatEVM2IstanbulCompatible"
```
```suggestion
		return "CodeFormatEVMIstanbulCompatible"
```

yoomee1313 replied:
thanks, reflected at https://github.com/klaytn/klaytn/pull/912/commits/94cfe30209ae84db4dc9fdfdbb376499c2de156c

aidan-kwon commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,31 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+	}
+}
+
+func (evm *EVM) GetPrecompiledContractMap(caller, address common.Address) map[common.Address]PrecompiledContract {
```
The second parameter `address` is not needed anymore.

yoomee1313 replied:
thanks, reflected at https://github.com/klaytn/klaytn/pull/912/commits/94cfe30209ae84db4dc9fdfdbb376499c2de156c

winnie-byun commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,31 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+func (evm *EVM) GetPrecompiledContractMap(caller common.Address) map[common.Address]PrecompiledContract {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		// The following code is to make vmLog(0x09), feePayer(0x0a), validateSender(0x0b) precompiled contract
+		// works after the istanbulIncompatible change
+		// if contract(caller) is deployed before istanbul, use the old precompiled contract set (use constantinople)
+		//      (gas price is exactly same as before)
```
```suggestion
		// if contract(caller) is deployed before istanbul, use the old precompiled contract set (use constantinople)
		//      (gas price policy also follows constantinople rules)
		// Without these lines, contracts that are deployed before istabul and uses vmLog(0x09), feePayer(0x0a), validateSender(0x0b) won't not work properly.
```

winnie-byun commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,31 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+		}
+		return PrecompiledContractsIstanbul
+	default:
+		return PrecompiledContractsConstantinople
```
Shouldn't this be the recent one?
```suggestion
		return PrecompiledContractsIstanbul
```

yoomee1313 replied:
PrecompiledContractsConstantinople is returned when evm.chainRules.IsIstanbul is false. It means that the IstanbulCompatible block number has not been reached yet. So it returns PrecompiledContractsIstanbul.

winnie-byun commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,30 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+	case evm.chainRules.IsIstanbul:
+		// if contract(caller) is deployed before istanbul, use the old precompiled contract set (use constantinople)
+		//      (gas price policy also follows constantinople rules)
+		// Without these lines, contracts that are deployed before istanbul and uses vmLog(0x09), feePayer(0x0a), validateSender(0x0b) won't not work properly.
```
```suggestion
		// Without these lines, contracts that are deployed before istanbul and uses vmLog(0x09), feePayer(0x0a), validateSender(0x0b) won't work properly.
```

yoomee1313 replied:
Reflected at 5121958

winnie-byun commented on params/protocol_params.go:
```diff
@@ -149,14 +149,17 @@ func GetMaximumExtraDataSize() uint64 {
 
 const (
-	CodeFormatEVM CodeFormat = iota
+	CodeFormatEVM                   CodeFormat = iota // contracts deployed before IstanbulCompatible are assigned CodeFormatEVM
```
```suggestion
	CodeFormatEVMConstantinople                   CodeFormat = iota // contracts deployed before IstanbulCompatible are assigned CodeFormatEVM
```

yoomee1313 replied:
https://github.com/klaytn/klaytn/commit/ac19b0238823c7c8e7982b89e4197415472dfc12 reflected

yoomee1313 replied:
This is reverted.

yoomee1313 commented:
Several new commits are updated.
* https://github.com/klaytn/klaytn/pull/912/commits/32be181d57818dc3c7345e28f55b75959bc9540c added unit test
* https://github.com/klaytn/klaytn/pull/912/commits/d8d6e5a03047dcf2b64d6c9be079e333a9b986cd comment revision(reflect review)
* https://github.com/klaytn/klaytn/pull/912/commits/4771cf4ad76f0fa265ad81468f71781467c81c7a remove currentCodeFormat function and add this logic into CreateSmartContract
* https://github.com/klaytn/klaytn/pull/912/commits/ac19b0238823c7c8e7982b89e4197415472dfc12 change name of params.CodeFormat(reflect review)
* https://github.com/klaytn/klaytn/pull/912/commits/e362d9e8ac99683266e6df20909bbaa6f42ce05b modify VM.Create, VM.CreateWithAddress since codeformat parameter is no longer needed
+ linter returned cognitive complexity error. i'll fix it.

yoomee1313 commented:
@winnie-byun @KimKyungup @aidan-kwon @ehnuje Several new commits are updated. Please take a look.

aidan-kwon commented on blockchain/vm/runtime/runtime.go:
```diff
@@ -107,7 +107,7 @@ func Execute(code, input []byte, cfg *Config) ([]byte, *state.StateDB, error) {
 		sender  = vm.AccountRef(cfg.Origin)
 	)
-	cfg.State.CreateSmartContractAccount(address, params.CodeFormatEVM)
+	cfg.State.CreateSmartContractAccount(address, true)
```
`CodeFormatEVMConstantinople` was assigned in the previous code, but your change assigns `CodeFormatEVMIstanbul`. Why did you change it?

yoomee1313 replied:
I reverted this code.

yoomee1313 replied:
@aidan-kwon I changed the code that i thought CodeFormat field is deprecated. I reverted this code.

aidan-kwon commented on blockchain/state/statedb.go:
```diff
@@ -326,6 +326,18 @@ func (self *StateDB) IsValidCodeFormat(addr common.Address) bool {
 	return false
 }
 
+func (self *StateDB) GetCodeFormat(addr common.Address) params.CodeFormat {
```
I think we need to add comments remarking that the method can return an invalid code format, `params.CodeFormatLast` and it can be validated with `CodeFormat.Validate()`.

yoomee1313 replied:
Thanks! Reflected at 512195885257c4ea89242bdbe39a76c626584a92

aidan-kwon commented:
@yoomee1313 Documentation is needed when a new version with this PR is released.
- Change `CodeFormatEVM` to `CodeFormatEVMConstantinople` in all contents of Klaytn Docs. Especially, the codeFormat section of https://docs.klaytn.com/klaytn/design/transactions/basic#txtypesmartcontractdeploy should be changed. 

Announcement is needed before the next incompatible change is applied. 
- Announce to Klaytn ecosystem tool (e.g., caver) developers and developer communities that the default codeFormat of legacy type txs will be changed from 0 indicating `CodeFormatEVMConstantinople` to 1 indicating `CodeFormatEVMIstanbul`. 
- Encourage Klaytn developer communities to use `CodeFormatEVMIstanbul` for contract deploying type txs after incompatible change.

yoomee1313 commented:
@aidan-kwon I reverted `CodeFormatEVMConstantinople` to `CodeFormatEVM`, and also removed `CodeFormatEVMIstanbul`.
In conclusion, I remove CodeFormatEVMIstanbul. Instead, first bit(MSB) of codeFormat is changed to a istanbulHF field.
case 1. CodeFormat [0 0 0 0 0 0 0 0]. MSB is 0, so the contract is deployed before istanbulHF.
case 2. CodeFormat [1 0 0 0 0 0 0 0]. MSB is 1, so the contract is deployed after istanbulHF.
cc. @winnie-byun @ehnuje

ehnuje commented on params/protocol_params.go:
```diff
@@ -163,8 +169,22 @@ func (t CodeFormat) String() string {
 	return false
 }
+
+func (t CodeFormat) IsIstanbulHF() bool {
```
Minor comment, however, how about considering `AfterInstanbulHF`? 
As "a CodeFormat is Instanbul HF" sounds a bit confusing. How do you think?

yoomee1313 replied:
Thanks! I'll change the name to `AfterIstanbulHF`.

yoomee1313 replied:
Changed the name to `IsDeployedAfterIstanbulHF` (https://github.com/klaytn/klaytn/commit/b62138adad3b89433c763458df317a578be8d84a)

ehnuje commented on blockchain/vm/evm_test.go:
```diff
@@ -0,0 +1,108 @@
+		ret, gas, err := vmenv.Call(AccountRef(callerAddr), common.HexToAddress(tt.addr), tt.input, math.MaxUint64, new(big.Int))
+
+		// compare with expected data
+		assert.Equal(t, tt.ret, common.Bytes2Hex(ret))
+		assert.Equal(t, tt.gas, math.MaxUint64-gas)
+		assert.Equal(t, tt.err, err)
```
I think you declared `TestField` and `ExpectedField` to categorize related values. However, if you do not give them  names, it is hard to understand where these fields come from. It is a matter of taste, but I think explicit naming helps readers understand the structure of the code, though we should type a bit more codes. 😅

yoomee1313 replied:
Thanks for your advice, please take a look https://github.com/klaytn/klaytn/pull/912/commits/b62138adad3b89433c763458df317a578be8d84a

ehnuje commented on blockchain/state/statedb.go:
```diff
@@ -326,6 +326,20 @@ func (self *StateDB) IsValidCodeFormat(addr common.Address) bool {
 	return false
 }
 
+// GetCodeFormat can return an invalid codeFormat, params.CodeFormatLast.
```
This sentence sounds a bit confusing. What did you intend?

yoomee1313 replied:
I wrote the comment to tell that this function can return invalid codeFormat if `getStateObject(addr)` or `GetProgramAccount(stateObject.account)` is failed. please take a look https://github.com/klaytn/klaytn/commit/b62138adad3b89433c763458df317a578be8d84a

yoomee1313 commented on params/protocol_params.go:
```diff
@@ -154,7 +160,7 @@ const (
 
 func (t CodeFormat) String() string {
-	switch t {
+	switch t & 0b01111111 { // IstanbulHF field is irrelevant with CodeFormat.
```
thanks, declared them as const variables https://github.com/klaytn/klaytn/commit/b62138adad3b89433c763458df317a578be8d84a

winnie-byun commented on params/protocol_params.go:
```diff
@@ -163,8 +176,24 @@ func (t CodeFormat) String() string {
+		return true
+	}
+}
+
```
```suggestion

// SetIstanbulHFField should be called for contracts deployed after Istanbul hard fork.
```

winnie-byun commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +539,21 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
 }
 
+func (evm *EVM) GetPrecompiledContractMap(caller common.Address) map[common.Address]PrecompiledContract {
+	switch {
+	case evm.chainRules.IsIstanbul:
+		// if contract(caller) is deployed before istanbul, use the old precompiled contract set (use constantinople)
+		//      (gas price policy also follows constantinople rules)
+		// Without these lines, contracts that are deployed before istanbul and uses 0x09-0x0b won't work properly.
+		if evm.StateDB.GetCodeFormat(caller).IsDeployedAfterIstanbulHF() {
+			return PrecompiledContractsIstanbul
+		}
+		return PrecompiledContractsConstantinople
+	default:
+		return PrecompiledContractsConstantinople
+	}
```
Doesn't this logic apply for after versions, too?
```
	// if contract(caller) is deployed before istanbul, use the old precompiled contract set (use constantinople)
	//      (gas price policy also follows constantinople rules)
	// Without these lines, contracts that are deployed before istanbul and uses 0x09-0x0b won't work properly.
	if !evm.StateDB.GetCodeFormat(caller).IsDeployedAfterIstanbulHF() {
			return PrecompiledContractsConstantinople
	}
	switch {
	case evm.chainRules.IsPetersburg: // this line is an example for next hard fork
		return PrecompiledContractsPetersburg
	case evm.chainRules.IsIstanbul:
		return PrecompiledContractsIstanbul
	default:
		return PrecompiledContractsConstantinople
	}
```

yoomee1313 replied:
@winnie-byun Thanks for picking up the unexpected part! Before, it does not consider another hardforks. Fixed the function at https://github.com/klaytn/klaytn/pull/912/commits/ad9898839583ea58277714efc58d0ae3012cc37c. Please take a look.

aidan-kwon commented on blockchain/vm/evm.go:
```diff
@@ -452,6 +452,7 @@ func (evm *EVM) create(caller types.ContractRef, codeAndHash *codeAndHash, gas u
 	// TODO-Klaytn-Accounts: for now, smart contract accounts cannot withdraw KLAYs via ValueTransfer
 	//   because the account key is set to AccountKeyFail by default.
 	//   Need to make a decision of the key type.
+	codeFormat.SetIstanbulHFField(evm.chainRules.IsIstanbul)
```
There are other places `CreateSmartContractAccountWithKey()` is called. Why don't you add `codeFormat.SetIstanbulHFField()` inside of `CreateSmartContractAccountWithKey()` ?

aidan-kwon commented on params/protocol_params.go:
```diff
@@ -146,15 +146,28 @@ func GetMaximumExtraDataSize() uint64 {
+// To Record the deployment time, the MSB of CodeFormat is changed to a istanbulHF field.
+// case 1. CodeFormat [0 0 0 0 0 0 0 0]. MSB is 0, so the contract is deployed before istanbulHF.
+// case 2. CodeFormat [1 0 0 0 0 0 0 0]. MSB is 1, so the contract is deployed after istanbulHF.
+
```
```suggestion
```

aidan-kwon commented on params/protocol_params.go:
```diff
@@ -146,15 +146,28 @@ func GetMaximumExtraDataSize() uint64 {
+
 const (
 	CodeFormatEVM CodeFormat = iota
 	CodeFormatLast
```
We need a comment ensuring that "CodeFormatLast should be smaller than CodeFormatFieldZeroCF".

yoomee1313 replied:
As we discussed on offline, size of the codeFormat field is fixed, 4 bit, so CodeFormatLast should be smaller than 16. Anyway, the comment which contains size of each field is added to the above of `type codeInfo uint8`( Naming is really difficult.. For now, I named it `codeInfo`).

aidan-kwon commented:
@yoomee1313 If a user retrieves his contract account info with `klay_GetAccount` API after generating it with codeformat 0, the returns of API may contain a different codeformat value, 128. 
If we intend to conceal the meaning of MSB of codeformat to users, we need to investigate and change all usage of codeformat.

KimKyungup commented on params/protocol_params.go:
```diff
@@ -146,28 +146,54 @@ func GetMaximumExtraDataSize() uint64 {
 const (
-	CodeFormatEVM                   CodeFormat = iota // contracts deployed before IstanbulCompatible are assigned CodeFormatEVM
-	CodeFormatEVMIstanbulCompatible                   // contracts deployed after IstanbulCompatible are assigned CodeFormatEVMIstanbulCompatible
+	IstanbulHFFieldZeroCF = 127 // 0b01111111: hardfork field value is zero, but codeFormat field values are one
+	CodeFormatFieldZeroCF = 128 // 0b10000000: hardfork field value is one, but codeFormat field values are zero
```
```suggestion
	IstanbulHFFieldZeroCF = 0b01111111 // hardfork field value is zero, but codeFormat field values are one
	CodeFormatFieldZeroCF = 0b10000000 // hardfork field value is one, but codeFormat field values are zero
```

yoomee1313 replied:
Good. I'll change it as you suggested.

yoomee1313 commented:
@aidan-kwon Thanks for the comments. I will fix it like that.

yoomee1313 commented on params/protocol_params.go:
```diff
@@ -146,28 +146,54 @@ func GetMaximumExtraDataSize() uint64 {
+// For this reason, contracts deployed before HF use old precompiledContract map so it can work even after HF.
+// To Record the deployment time, the MSB of CodeFormat is changed to a istanbulHF field.
+// case 1. CodeFormat [0 0 0 0 0 0 0 0]. MSB is 0, so the contract is deployed before istanbulHF.
+// case 2. CodeFormat [1 0 0 0 0 0 0 0]. MSB is 1, so the contract is deployed after istanbulHF.
```
Agreed. BTW, the size is no problem. Since it is increased from the front, it can be increased easily if necessary.

aidan-kwon replied:
Good suggestion. Or we can reserve 3 Bytes for later use. It's hard for me to think the later use can be categorized in the same subject with this MSB.

winnie-byun commented on params/protocol_params.go:
```diff
@@ -168,3 +190,29 @@ func (t CodeFormat) Validate() bool {
 	return false
 }
+
+func (t CodeFormat) GenerateCodeInfo(r Rules) CodeInfo {
+	return CodeInfo(t) | CodeInfo(GenerateVmVersion(r))
+}
```
How about changing the function name and add a function like the following?
```suggestion
func (t CodeFormat) CodeInfoFromRules(r Rules) CodeInfo {
	return CodeInfo(t) | CodeInfo(GenerateVmVersion(r))
}
func CodeInfoFromCodeFormat(c CodeFormat) CodeInfo {
	return CodeInfo(c)
}
func CodeInfoFromVmVersion(c CodeFormat) CodeInfo {
	return CodeInfo(c)
}
```
We can set like `AccountValueKeyCodeInfo:      params.CodeInfoFromCodeFormat(params.CodeFormatEVM),` in account_test.go
and `sca.codeInfo = params.CodeInfoFromCodeFormat(serialized.CodeFormat) | params.CodeInfoFromVmVersion(serialized.VmVersion)` in smart_contract_account.go

yoomee1313 replied:
Forgot to response at this comment.. I reflected at https://github.com/klaytn/klaytn/pull/912/commits/afadf16289832acbd0491d966c9253ddbd8f8861!

winnie-byun commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,35 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
+	// There are contracts which uses latest precompiled contract map (regardless of deployment time)
+	// If new HF is added, please add new case below
+	switch {
+	case evm.chainRules.IsIstanbul:
+		return PrecompiledContractsIstanbul
```
Shouldn't we add `IsConstantinople` in Rules?

yoomee1313 replied:
@winnie-byun Can I ask why this field is needed? In here, default case means use PrecompiledContractsConstantinople, not a Constantinople hard fork.

winnie-byun replied:
For the next vm version, we have to make `IsConstantinople`. It seems awkward to me if we make something for the previous version.
It is not critical problem.

yoomee1313 replied:
`IsConstantinople` is not needed. If new vm version is added, add a new case statement as shown in the code below.
```
switch contractVmVersion {
case VmVersionIstanbul:
        return PrecompiledContractsBerlin (last hard fork of that vm version)
default:
        return PrecompiledContractsConstantinople
```

ehnuje reviewed (dismissed):
LGTM except for others' comments

yoomee1313 commented:
@KimKyungup Please take a look.

KimKyungup commented on params/protocol_params.go:
```diff
@@ -168,3 +198,29 @@ func (t CodeFormat) Validate() bool {
+}
+
+// Supporting VmVersion
+type VmVersion CodeInfo
```
Why are VmVersion and CodeFormat same `CodeInfo` type?
How about uint8?

yoomee1313 replied:
Thanks for the comment :) I reflected at https://github.com/klaytn/klaytn/pull/912/commits/f21a7293d0e66c538653f3713661fa1c0d7a6a49

KimKyungup commented on params/protocol_params.go:
```diff
@@ -146,13 +146,43 @@ func GetMaximumExtraDataSize() uint64 {
+}
+
+// Supporting CodeFormat
+type CodeFormat CodeInfo
```
In 166 line, `CodeFormat` is used.
So please move this declaration to the line before that.

yoomee1313 replied:
`GetCodeFormat` belongs to the codeInfo, so I located the `GetCodeFormat` and `GetVmVersion` at line 166.

KimKyungup commented on params/protocol_params.go:
```diff
@@ -146,13 +146,43 @@ func GetMaximumExtraDataSize() uint64 {
+}
+
+func (t CodeInfo) GetVmVersion() VmVersion {
+	return VmVersion(t & vmVersionBitMask)
```
How about using shift operator?
I think 0x01 is better than 0x10 for version 1.

```suggestion
	return VmVersion(t & vmVersionBitMask >> 4)
```

yoomee1313 replied:
Thanks for the comment! I changed it as you suggested. https://github.com/klaytn/klaytn/pull/912/commits/f21a7293d0e66c538653f3713661fa1c0d7a6a49

KimKyungup commented on params/protocol_params.go:
```diff
@@ -168,3 +198,29 @@ func (t CodeFormat) Validate() bool {
+type VmVersion CodeInfo
+
+const (
+	VmVersionConstantinople VmVersion = 0b00000000 // Deployed at Constantinople
```
```suggestion
	VmVersionConstantinople VmVersion = 0 // Deployed at Constantinople
```

yoomee1313 replied:
I changed it to use `iota`.

KimKyungup commented on params/protocol_params.go:
```diff
@@ -168,3 +198,29 @@ func (t CodeFormat) Validate() bool {
+
+const (
+	VmVersionConstantinople VmVersion = 0b00000000 // Deployed at Constantinople
+	VmVersionIstanbul                 = 0b00010000 // Deployed at Istanbul, ...(later HFs would be added)
```
```suggestion
	VmVersionIstanbul                 = 1 // Deployed at Istanbul, ...(later HFs would be added)
```

yoomee1313 replied:
I changed it to use `iota`.

KimKyungup commented on blockchain/vm/evm.go:
```diff
@@ -538,6 +538,35 @@ func (evm *EVM) CreateWithAddress(caller types.ContractRef, code []byte, gas uin
 }
 
+func (evm *EVM) GetPrecompiledContractMap(caller common.Address) map[common.Address]PrecompiledContract {
+	if codeInfo, ok := evm.StateDB.GetCodeInfo(caller); ok {
```
I think we need to consider chain rule first.
And then if it is istanbul chain rule, the vm version should be considered.

Therefore how about this code?
```suggestion
	switch {
	case evm.chainRules.IsIstanbul:
		codeInfo := evm.StateDB.GetCodeInfo(caller)
		if codeInfo ==  VmVersionIstanbul {
			return PrecompiledContractsIstanbul
		}
		return PrecompiledContractsConstantinople
	default:
		return PrecompiledContractsConstantinople	
```

yoomee1313 replied:
First of all, I put the `GetCodeInfo` line at first because this function is not only used for smart contract addresses, but also with zero address, etc.
Istanbul chain rule info is used when allocating vmVersion info to the `currentBlockVMVersion` variable. BTW, I tried to extract the vmVersion logic from the chain rules switch because vmVersion there may be a difference of 2 or more.

yoomee1313 replied:
@KimKyungup I refactor the code as you suggested and also changed the naming of VmVersion. 
change naming of VmVersion https://github.com/klaytn/klaytn/pull/912/commits/afa456dfc109d315a894d5e29840a31f43598f3e
refactor GetPrecompiledContractMap https://github.com/klaytn/klaytn/pull/912/commits/db60f58712f54075d16abb4cfea97a8ef893bf0d

KimKyungup commented on params/protocol_params.go:
```diff
@@ -168,3 +198,29 @@ func (t CodeFormat) Validate() bool {
+func GenerateVmVersion(r Rules) VmVersion {
+	switch {
+	// If new HF is added, please add new case below
+	case r.IsIstanbul:
```
```suggestion
//	case.r.IsNextHF:          // If this HF is backward compatible with vmVersion1.
	case r.IsIstanbul:
```

winnie-byun reviewed (approved):
LGTM

yoomee1313 commented:
@aidan-kwon CodeInfo part is refactored. Please take a look.

aidan-kwon commented on params/protocol_params.go:
```diff
@@ -162,9 +172,61 @@ func (t CodeFormat) String() string {
+)
+
+func NewCodeInfo(codeFormat CodeFormat, vmVersion VmVersion) CodeInfo {
+	return CodeInfo(codeFormat) | CodeInfo(vmVersion)<<4
```
It's not necessary but I think masking for codeFormat is more safe wasy
```suggestion
	return CodeInfo(codeFormat & codeFormatBitMask) | CodeInfo(vmVersion)<<4
```

yoomee1313 replied:
Agreed that it is not necessary for now, but it is more safe.. thanks for the suggestion!

aidan-kwon commented on params/protocol_params.go:
```diff
@@ -162,9 +172,61 @@ func (t CodeFormat) String() string {
+		vmVersion = VmVersion0
 	}
-	return false
+	return CodeInfo(codeFormat) | CodeInfo(vmVersion)<<4
```
```suggestion
	return NewCodeInfo(codeFormat, vmVersion)
```

aidan-kwon commented on params/protocol_params.go:
```diff
@@ -146,13 +147,22 @@ func GetMaximumExtraDataSize() uint64 {
+// CodeFormat is the version of the interpreter that smart contract uses
 type CodeFormat uint8
 
+// Supporting CodeFormat
```
```suggestion
// Supporting CodeFormat
// CodeFormatLast should be equal or less than 16 because only the last 4 bits of CodeFormat are used for CodeInfo.
```

aidan-kwon reviewed (dismissed):
LGTM except for minors

KimKyungup commented on blockchain/types/account/account_test.go:
```diff
@@ -149,7 +149,7 @@ func genSCA() *SmartContractAccount {
 		AccountValueKeyStorageRoot:   genRandomHash(),
 		AccountValueKeyCodeHash:      genRandomHash().Bytes(),
-		AccountValueKeyCodeInfo:      params.CodeFormatEVM.CodeInfoFromCodeFormat(),
+		AccountValueKeyCodeInfo:      params.CodeInfo(0),
```
`params.CodeInfo(0)` is same with `NewCodeInfo(VmVersion0,CodeFormatEVM)`?

yoomee1313 replied:
@KimKyungup yes, it's same.
